lib/ble.toit

```diff
@@ -257,23 +270,157 @@ class Client:
     state := resource_state_.wait_for_state STARTED_EVENT_
     resource_state_.clear_state STARTED_EVENT_
 
+class ServerConfig:
+  resource_group_ := null
+  services/List := []
+
+  constructor:
+    resource_group_ = ble_server_config_init_
+
+  add_service uuid -> Service:
+    service := Service resource_group_ uuid
+    services.add service
+    return Service resource_group_ uuid
+
+
+/** 
+Defines a BLE service with characteristics 
+*/
+class Service:
+  /**
+  The UUID of the service.
+  
+  For 16 and 32 bit UUIDs, form the BLE variant with the top-level uuid function.
+  */
+  uuid/uuid_pkg.Uuid
+  
+  resource_group_/any
+  resource_/any
+
+  characteristics_/Map := {:}
+
+  constructor .resource_group_ .uuid:
+    resource_ = ble_add_server_service_ resource_group_ uuid.to_byte_array
+
+  add_read_only_characteristic uuid --value=#[]-> ReadOnlyCharacteristic:
+    return ReadOnlyCharacteristic this uuid value
+
+  add_write_only_characteristic uuid -> WriteOnlyCharacteristic:
+    return WriteOnlyCharacteristic this uuid
+
+  add_read_write_characteristic uuid --value=#[]-> ReadWriteCharacteristic:
+    return ReadWriteCharacteristic this uuid value
+
+  add_notification_characteristic uuid -> NotificationCharacteristic:
+    return NotificationCharacteristic this uuid
+
+  get_characteristic uuid -> Characteristic?:
+    return characteristics_[uuid]
+
+
+BLE_CHR_TYPE_READ_ONLY_    ::= 1
+BLE_CHR_TYPE_WRITE_ONLY_   ::= 2
+BLE_CHR_TYPE_READ_WRITE_   ::= 3
+BLE_CHR_TYPE_NOTIFICATION_ ::= 4
+
+BLE_WAIT_RECV_             ::= 1 << 0
+BLE_WAIT_ACCESSED_         ::= 1 << 1
+BLE_WAIT_SUBSCRIBED_       ::= 1 << 2
+
+abstract class Characteristic:
+  /**
+  The UUID of the characteristic.
+  */
+  uuid/uuid_pkg.Uuid
+
+  state_/ResourceState_
+
+  constructor service/Service resource .uuid:
+    state_ = ResourceState_ service.resource_group_ resource
```

Comment:
Is this where you would want to register the `Characteristic` in the `service.characteristics_` map?